File: android/feature/aiconnect/src/main/java/com/enterprise/feature/aiconnect/domain/usecases/ChatUseCases.kt
package com.enterprise.feature.aiconnect.domain.usecases

import com.enterprise.core.domain.entities.DomainResult
import com.enterprise.core.domain.entities.UCPMessage
import com.enterprise.core.domain.repositories.ChatRepository
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

/** Sends a message via chat repository. */
class SendMessageUseCase @Inject constructor(
    private val chatRepository: ChatRepository
) {
    suspend operator fun invoke(text: String): DomainResult<UCPMessage> {
        return chatRepository.sendMessage(text)
    }
}

/** Observes incoming chat message stream. */
class ObserveChatStreamUseCase @Inject constructor(
    private val chatRepository: ChatRepository
) {
    operator fun invoke(): Flow<UCPMessage> {
        return chatRepository.messageStream()
    }
}

/** Gets chat history from server. */
class GetChatHistoryUseCase @Inject constructor(
    private val chatRepository: ChatRepository
) {
    suspend operator fun invoke(limit: Int = 50): DomainResult<List<UCPMessage>> {
        return chatRepository.getHistory(limit)
    }
}

File: ios/EnterpriseMobileApp/Sources/Feature/AIConnect/Domain/UseCases/ChatUseCases.swift
import Foundation
import CoreDomain

/// Use case for sending messages and receiving AI responses.
public struct SendMessageUseCase: Sendable {
    private let chatRepository: ChatRepository

    public init(chatRepository: ChatRepository) {
        self.chatRepository = chatRepository
    }

    public func execute(_ text: String) async -> DomainResult<UCPMessage> {
        await chatRepository.sendMessage(text)
    }
}

/// Use case for observing incoming chat messages via WebSocket.
public struct ObserveChatStreamUseCase: Sendable {
    private let chatRepository: ChatRepository

    public init(chatRepository: ChatRepository) {
        self.chatRepository = chatRepository
    }

    public func execute() -> AsyncStream<UCPMessage> {
        chatRepository.messageStream()
    }
}

/// Use case for loading chat history.
public struct GetChatHistoryUseCase: Sendable {
    private let chatRepository: ChatRepository

    public init(chatRepository: ChatRepository) {
        self.chatRepository = chatRepository
    }

    public func execute(limit: Int = 50) async -> DomainResult<[UCPMessage]> {
        await chatRepository.getHistory(limit: limit)
    }
}
